### PR TITLE
MPI support

### PR DESCRIPTION
**Pull request by vchuravy**
- add MPI routines
- add rudimentary MPI support
- fixup! add MPI routines

### CODE REVIEW

**Comment by vchuravy** [comment]
Will need a released version of MPI.jl.

**Comment by eschnett** [comment]
@vchuravy I don't quite understand what you're trying to do. I assume you intend the following setup:
- Build LAMMPS against MPItrampoline_jll
- Tell Julia to use MPICH_jll
- At run time, tell MPItrampoline to use the MPICH from MPICH_jll

I don't think this would work.

However: MPItrampoline_jll has a feature to make it easier to use it: It ships with a wrapped MPICH! (For technical reasons this can't be the MPICH from MPICH_jll, having to do with how BinaryBuilder loads packages.) This makes it very easy to get started with MPItrampoline_jll: Just use it, and you'll be using an automatically installed wrapped MPICH so that everything will work out of the box. You don't need a system MPI. (Yggdrasil configures MPItrampoline to have a fallback setting for a wrapped MPI.)

I suggest the following instead:
- Build LAMMPS against MPItrampoline_jll
- Tell Julia to use MPItrampoline_jll: set `.julia/prefs/MPI.toml` to `binary = "MPItrampoline_jll"` and then `build MPI`

Nothing else should be necessary.



Long explanation why mixing MPItrampoline_jll and MPICH_jll is not a good idea: LAMMPS would pull in an `MPI_Send` from MPItrampoline_jll, and Julia would load an `MPI_Send` from MPICH_jll, and these two are not compatible. You would need to be very careful that each package resolves symbols against the right MPI library. Furthermore, you would be mixing incompatible types for `MPI_Comm`. Passing communicators or other MPI objects between Julia and LAMMPS would segfault. There should be only one MPI implementation active in an application.

**Comment by eschnett** [comment]
At a quick glance your changes look correct. I don't understand where `mpi_header_dir` would come from, though. Is `MPI.jl` expected to export it?

File: src/LAMMPS.jl [hunk 1]
```diff
@@ -18,7 +18,11 @@ mutable struct LMP
         end
 
         GC.@preserve args begin
-            handle = API.lammps_open_no_mpi(length(args), argsv, C_NULL)
+            if comm !== nothing
+                handle = API.lammps_open(length(args), argsv, comm, C_NULL)
```
Comment (pabloferz) [inline src/LAMMPS.jl:22]:
```suggestion
                if !MPI.Initialized()
                    error("MPI has not been initialized. Make sure to call `MPI.Init()`"
                end
                handle = API.lammps_open(length(args), argsv, comm, C_NULL)
```

Comment (pabloferz) [reply src/LAMMPS.jl:22]: I think this might segfault otherwise.

**Comment by vchuravy** [comment]
@eschnett 

> Build LAMMPS against MPItrampoline_jll

That is already done. Yeah my concern is the other side. How do we get things to work out-of-the box in the ecosystem.
My worry is someone setting OpenMPI_jll, MPICH_jll, or a system MPI as the MPI.jl backend and then getting inconsistent state and hard to debug errors.

> I don't understand where mpi_header_dir would come from, though.
From MPI library that was LAMMPS was built against in this case MPItrampoline_jll

File: src/LAMMPS.jl [hunk 2]
```diff
@@ -18,7 +18,14 @@ mutable struct LMP
         end
 
         GC.@preserve args begin
-            handle = API.lammps_open_no_mpi(length(args), argsv, C_NULL)
+            if comm !== nothing
+                if !MPI.Initialized()
+                    error("MPI has not been initialized. Make sure to first call `MPI.Init()`"
```
Comment (vchuravy) [inline src/LAMMPS.jl:23]:
```suggestion
                    error("MPI has not been initialized. Make sure to first call `MPI.Init()`")
```

**Comment by eschnett** [comment]
@vchuravy Using different MPI backends in MPI.jl and in Yggdrasil doesn't work now, and won't work in the future. If people want to use Yggdrasil-built packages together with MPI.jl, then all packages need to be built agains the same MPI implementation.

In other words, allowing different backends in MPI.jl should be disallowed in the future. Instead, people need to use MPItrampoline's mechanism. We should then change MPI.jl's `build` mechanism to be able to record people's preferences for MPItrampoline (by setting environment variables or similar).

**Comment by vchuravy** [comment]
I think we can have our cake and eat it to. We have this problem in multiple places, CUDA is another case. 

https://github.com/JuliaPackaging/Yggdrasil/pull/4073 and the PR it's builds upon should allow us to build things against MPICH, MPItrampoline and MSMPI. 

1. This has the benefit that switching the default in MPI.jl won't break the binary packages
2. We can use MPItrampoline to support people's system MPI (and probably should force that)

**Pull request merged**